### PR TITLE
fix widget garbage collection

### PR DESCRIPTION
**Pull request by akaszynski**
While working on #2547, it was discovered that widgets aren't being garbage collected correctly, leading to a segfault when rendering an example.

This PR moves `test_widgets.py` to take advantage of `check_gc` garbage collection fixture and verifys garbage collection of our widgets.

You can reproduce this issue (on Linux) on `main` with `vtk==9.1.0` with:
```py
import pyvista as pv
mesh = pv.ParametricConicSpiral()
pl = pv.Plotter()
_ = pl.add_mesh_clip_box(mesh, color='white')
pl.show()
```

### CODE REVIEW

File: pyvista/plotting/widgets.py [hunk 1]
```diff
@@ -140,10 +138,11 @@ def _the_callback(box_widget, event_id):
 
     def clear_box_widgets(self):
         """Disable all of the box widgets."""
-        if hasattr(self, 'box_widgets'):
-            for widget in self.box_widgets:
-                widget.Off()
-            del self.box_widgets
+        # this is necessary for garbage collection
+        for widget in list(self.box_widgets):
```
Comment (MatthewFlamm) [inline pyvista/plotting/widgets.py:142]: Given that self.box_widgets has type `List[object]`, why is wrapping this in `list` required?

**Review by MatthewFlamm — COMMENTED** [review]
I tested this and it works.

File: pyvista/plotting/widgets.py (hunk 1 above)
Comment (adeak) [reply pyvista/plotting/widgets.py:142]: Because it creates a copy that can be iterated over meanwhile deleting values from the original list.

What I don't understand is why we can't just do `sefl.box_widgets.clear()`.

File: pyvista/plotting/widgets.py [hunk 2]
```diff
@@ -140,10 +138,11 @@ def _the_callback(box_widget, event_id):
...
+            self.box_widgets.remove(widget)
+
+        self.box_widgets = []
```
Comment (adeak) [inline pyvista/plotting/widgets.py:145]: There is also a subtlety here. This might actually be related to the garbage collection bug.

As it is now, `box_widgets` and `_camera_widgets` are class (rather than instance) attributes. When you do `self.box_widgets.append(foo)` you are mutating `WidgetHelper.box_widgets`. All `WidgetHelper`s share this list, because it's tied to the type.

But when you do
```py
self.box_widgets = []
```
you create an _instance_ attribute that shadows the class attribute. From that point on this `WidgetHelper` (this `Plotter`) will have a dedicated set of box widgets.

So two points here:
  1. If we go with my comment that directly clears `self.box_widgets` we circumvent this problem by always mutating the class attribute.
  2. Are these two class attributes on purpose? Shouldn't these widgets be handled on a per-plotter level? This would also mean that a `Plotter`'s demise involves clearing these lists.

File: pyvista/plotting/widgets.py (hunk 1 above)
Comment (akaszynski) [reply pyvista/plotting/widgets.py:142]: > What I don't understand is why we can't just do self.box_widgets.clear(). Or self.box_widgets[:] = [] (to replace the whole function with).

I have no idea. It should have garbage collected once the last reference disappeared by overwriting the list. Through trial and error this was the only solution that worked.

File: pyvista/plotting/widgets.py (hunk 2 above)
Comment (akaszynski) [reply pyvista/plotting/widgets.py:145]: > There is also a subtlety here. This might actually be related to the garbage collection bug.*

That must be it! Let me fix that.

File: pyvista/plotting/widgets.py (hunk 1 above)
Comment (adeak) [reply pyvista/plotting/widgets.py]: For completeness' sake: we could, if we remove `self.box_widgets = []` from a few lines lower and always use the class attribute :)

**Comment by akaszynski** [comment]
> Are these two lists class attributes on purpose? Shouldn't these widgets be handled on a per-plotter level? This would also mean that a Plotter's demise involves clearing these lists. To be clear this would involve creating these lists in __init__ (and only there).

Done in 48736a7. Moved all widgets to `BasePlotter.__init__`. This cleans up collection and fixes our anti-pattern of checking for attributes and adding them at run time rather than having them in `__init__`.

**Comment by akaszynski** [comment]
> I'd go for .clear() anyway for semantics

Done in 70c62fb.

File: pyvista/plotting/plotting.py [hunk 3]
```diff
@@ -278,6 +278,22 @@ def __init__(
...
+        self.plane_clipped_meshes = []
+        self.plane_sliced_meshes = []
+        self.line_widgets = []
+        self.slider_widgets = []
+        self.threshold_meshes = []
+        self.isovalue_meshes = []
+        self.spline_widgets = []
+        self.spline_sliced_meshes = []
+        self.sphere_widgets = []
+        self.button_widgets = []
```
Comment (adeak) [inline pyvista/plotting/plotting.py:295]: Ideally we should move all these into `WidgetHelper.__init__`. But for that we would have to take care to call `super().__init__` in our own `__init__`s.

Here's the method resolution order for `Plotter`:
```py
>>> pv.Plotter.mro()
[<class 'pyvista.plotting.plotting.Plotter'>, <class 'pyvista.plotting.plotting.BasePlotter'>, <class 'pyvista.plotting.picking.PickingHelper'>, <class 'pyvista.plotting.widgets.WidgetHelper'>, <class 'object'>]
```
We'd definitely have to call `super().__init__()` in `BasePlotter.__init__()`, but that might be enough since `PickingHelper` has no `__init__()` of its own. And the only reason for this would be purity; not sure it's worth it.

Then again it might be good form to prepare `BasePlotter` for cooperative multiple inheritance if it's easy to do. Because if someone does
```py
class DuperPlotter(BasePlotter, DuperType):
    ...
```
then `BasePlotter.__init__()` would be called first as per the MRO, and `DuperType.__init__()` wouldn't have a chance to run unless we call `super().__init__` in `BasePlotter`.


_However_, one thing that could be tricky is deciding what args to pass on to `super().__init__()`. Because we might want to pass on args not caught by us, but we also must prepare for the possibility that `super()` resolves to `object` which doesn't take any parameters in its `__init__()`. It might work to add `*args, **kwargs` to our `__init__()` signatures and pass those on... when everything Just Works then we'd be passing on empty `args` and `kwargs` to `object.__init__()`.

File: pyvista/plotting/plotting.py [hunk 4]
```diff
@@ -278,6 +278,22 @@ def __init__(
         if self.theme.antialiasing:
             self.enable_anti_aliasing()
 
+        # widgets
+        self.camera_widgets = []
+        self.box_widgets = []
+        self.box_clipped_meshes = []
```
Comment (adeak) [inline pyvista/plotting/plotting.py:284]: And a concern that might be for another PR: is it on purpose that `self.plane_sliced_meshes` etc. mesh lists are never cleared, only appended to? When we disable a widget, do we want to keep these meshes around? Or should we clear the mesh lists as well?

File: pyvista/plotting/plotting.py (hunk 3 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py:295]: Looks like we've been using class attributes to avoid multiple inheritance, and to fix this I've done as you've suggested; moved the applicable to the `__init__`s of the subclasses. Not easy to wrap my head around, but learned a bit about [MRO](https://www.educative.io/edpresso/what-is-mro-in-python).

For anyone who wants to read up on multiple cooperative inheritance:
https://stackoverflow.com/a/27134600

Comment (adeak) [reply pyvista/plotting/plotting.py]: There's also Raymond Hettinger's [Python's `super()` considered super!](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/) as a great resource.

File: pyvista/plotting/plotting.py (hunk 4 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py]: I think it's ok to keep the plane sliced meshes since we're only disabling the widget, but we probably should make not of it in the docstring.

**Comment by akaszynski** [comment]
In [192932f](https://github.com/pyvista/pyvista/pull/2569/commits/192932fe454b78330a19e5bed020fd70b68319fc), I've also removed `Off` before removing the widget. On my env, this causes the plotter to fail to close for even the most trivial of scripts:

```py
import pyvista as pv
mesh = pv.ParametricConicSpiral()
pl = pv.Plotter()
_ = pl.add_mesh_clip_box(mesh, color='white')
pl.show()
```

Since we're removing the widget, there's no reason to disable it first.

File: pyvista/plotting/plotting.py [hunk 5]
```diff
@@ -197,6 +197,7 @@ def __init__(
         theme=None,
     ):
         """Initialize base plotter."""
+        super().__init__()  # cooperative multiple inheritance
```
Comment (adeak) [inline pyvista/plotting/plotting.py:200]: So for this to work I think we need to add `**kwargs` in the signature of our `__init__` (we probably don't have/want to have `*args` considering that everything we have are intended to be kwargs). And then
```suggestion
        theme=None,
        **kwargs,
    ):
        """Initialize base plotter."""
        super().__init__(**kwargs)  # cooperative multiple inheritance
```

Otherwise we're robbing `DuperType` the possibility to use additional kwargs in its `__init__`.

**Comment by adeak** [comment]
Since this PR doesn't have an "unrelated addition" slapped onto it: do we really need [`WidgetHelper` to be documented](https://dev.pyvista.org/api/plotting/_autosummary/pyvista.WidgetHelper.html?highlight=widgethelper#pyvista.WidgetHelper)? All the methods should be documented through the concrete subclasses, i.e. `Plotter`.

**Comment by akaszynski** [comment]
Calling this sufficiently reviewed. Thanks for your attention to detail @adeak!
